pkg/services/ngalert/state/testing.go

```diff
@@ -50,5 +48,5 @@ func (f *FakeRuleReader) ListAlertRules(_ context.Context, q *models.ListAlertRu
 
 type FakeHistorian struct{}
 
-func (f *FakeHistorian) RecordState(ctx context.Context, rule *models.AlertRule, labels data.Labels, evaluatedAt time.Time, currentData, previousData InstanceStateAndReason) {
+func (f *FakeHistorian) RecordStates(ctx context.Context, states []ContextualState) {
```

Comment:
I think keeping the rule would be more effective at least until we use RecordState to handle states by a specific rule

Reply:
I think adding pointer to rule to each `StateContext` captures this? I don't want to make this interface obviously "per-rule" if we can possibly avoid it, taking the flat list of transitions is the most flexible and clean API here IMO.

Reply:
If we copy the needed fields out of the rule before starting goroutine it shouldn't introduce any issues even if the rule changes underneath the state

Reply:
I think StateTransition would look cleaner with just current and old state.

Reply:
Done